File: IO/src/projects/s&p_493.r
#%%
# Load necessary libraries
library(dplyr)
library(ggplot2)
library(lubridate)
# Load necessary libraries
library(rvest)
library(dplyr)
library(lubridate)
library(quantmod)
library(tidyverse)

#%%
#%%
# Scrape S&P 500 companies from Wikipedia
url <- "https://en.wikipedia.org/wiki/List_of_S%26P_500_companies"
webpage <- read_html(url)
table <- html_nodes(webpage, "table")[[1]]
tickers <- html_text(html_nodes(table, "td:nth-child(1)"))

# Remove newline characters
tickers <- gsub("\n", "", tickers)
#%%
#%%
# Define start and end dates
start <- as.Date("2000-01-01")
end <- as.Date("2023-11-09")

# Download stock data
data <- do.call(cbind, lapply(tickers, function(ticker) {
    tryCatch(
        {
            getSymbols(ticker, src = "yahoo", from = start, to = end, auto.assign = FALSE)
        },
        error = function(e) {
            NULL
        }
    )
}))


# Convert to data frame and reshape
df <- data.frame(Date = index(data), coredata(data)) %>%
    pivot_longer(
        cols = -Date,
        names_to = c("Symbol", ".value"),
        names_pattern = "(.*)\\.(.*)"
    ) %>%
    arrange(Symbol, Date)

#%%
#%%

# Take tickers of top 7 tech giants in s&p 500
top_tickers <- c("AAPL", "MSFT", "AMZN", "NVDA", "GOOGL", "META", "TSLA")

# Create new columns 'type' and 'adj_close' based on conditions
df <- df %>%
    mutate(
        type = ifelse(Symbol %in% top_tickers, "Tech-7", "Others"),
    )

# Group by 'type' and 'Date', and calculate average 'adj_close'
df_grouped <- df %>%
    group_by(type, Date) %>%
    summarise(mean_close = mean(Close, weights=NULL, na.rm = TRUE)) %>%
    filter (Date >= as.Date("2020-01-01")) %>%
    mutate(price_index = mean_close / mean_close[1] * 100)

# Plot
ggplot(df_grouped, aes(x = Date, y = price_index, color = type)) +
    geom_line()
#%%
#%%

File: IO/src/projects/s&p_493.py
# %%
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import bs4 as bs
import requests
import yfinance as yf
from datetime import datetime
import duckdb

sql = lambda q: duckdb.sql(q).df()
sns.set()

# %%
resp = requests.get("http://en.wikipedia.org/wiki/List_of_S%26P_500_companies")
soup = bs.BeautifulSoup(resp.text, "lxml")
table = soup.find("table", {"class": "wikitable sortable"})

tickers = []

for row in table.findAll("tr")[1:]:
    ticker = row.findAll("td")[0].text
    tickers.append(ticker)


print(tickers)

tickers = [s.replace("\n", "") for s in tickers]
start = datetime(2000, 1, 1)
end = datetime(2023, 11, 9)
data = yf.download(tickers, start=start, end=end)

df = (
    data.stack()
    .reset_index()
    .rename(index=str, columns={"level_1": "Symbol"})
    .sort_values(["Symbol", "Date"])
)

# %%
# Take tickers of top 7 tech giants in s&p 500
top_tickers = ("AAPL", "MSFT", "AMZN", "GOOG", "GOOGL", "FB", "TSLA")
sns.lineplot(

    sql(
        f"""with grouped as (select *,
        case when Symbol in {top_tickers} then 'TG' else 'Other' end as type,
        case when Symbol in {top_tickers} then Close else Close/2 end as adj_close
        from df
        )
        select avg(adj_close) as adj_close,
                Date,
                type
        from grouped
        group by type, Date
        """
    ),
    y="adj_close",
    x="Date",
    hue="type",
)
plt.xlim(datetime(2018, 1, 1), datetime(2023, 12, 10))
